### DISCUSSION TITLE
Is there any way to force Cmake to install glfw even when it's already installed somewhere else?

### DISCUSSION THREAD

**Discussion by Lacusch** [codam-coding-college/MLX42#83] [Q&A] [answered]
When I install the MLX42 from a sub-module it's really conveniently installs glfw it it's not installed already. That means that I could to it in my Makefile so I don't depend on system install.

My Problem is that it glfw is already installed somewhere, then it doesn't download it again, so my Makefile is already looking for a non existing directory.

 In our Campus (42Heilbronn) we don't have glfw installed by default, so we need to install it with brew and it's location is in non standard places depending on the install.

So for me it would be really convenient if I could give CMake an option to always make glfw, so that in turn I don't need to hunt for it's specific location on other MAC-s.

PS: I hope that this ramble makes sense and thanks for you're work on the library it's a pleasure to use and the CMake version is easier to set up than ever.

**Comment by W2Wizard** [comment, marked as answer]
You could try and compile it with `cmake -B build -Dglfw_FOUND=0` if you want it to build and always download GLFW. Tho it might not work. You could create a fork and then modify the `CMakeLists.txt` and remove the if statement and `find_package` statement. That way it will clone and build it.

My real suggestion would be to tell your sysadmin to install it. CMake and GLFW are useful for advanced (specialization projects) later on anyway so why not include it?
If you're on MacOS he/she can ask for the Munki files I suppose.

**Reply by Lacusch** [reply]
Hey thanks for the quick answer. Compiling it with `cmake -B build -Dglfw_FOUND=0` doesn't solve my problem because cmake overrides the variable each time. But I found a small fix that can make it work.

```dif
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 8f7550b..69daa0f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -119,7 +119,9 @@ target_include_directories(mlx42 PUBLIC ${CMAKE_CURRENT_SOURCE_DIR}/include)
 # Dependencies
 # -----------------------------------------------------------------------------
 
+if(NOT DEFINED glfw_FOUND)
 find_package(glfw3)
+endif()
 if(NOT glfw3_FOUND)
        message("-- Please wait, fetching GLFW ...")
        include(${CMAKE_DIR}/LinkGLFW.cmake)
```

As far as I understand it this makes sure that find_package only runs it glfw_FOUND is not manually defined.
If you wouldn't mind accept the change I'm more than happy to create a pull request (It's basically only 2 lines anyway).

**Reply by Lacusch** [reply]
I made a draft pull request that you can check out #84

**Reply by W2Wizard** [reply]
Thats a possibility 👍 
I would very much welcome pull request however I suggest using variables instead:

```cmake
# Variables
# -----------------------------------------------------------------------------

set(GLFW_FETCH	ON CACHE BOOL "Clone and install GLFW")
set(DEBUG	OFF CACHE BOOL "Build MLX42 with debug flags and assertions")
```

That way if you for instance use the GUI of CMake you can more easily configure it:
<img width="620" alt="image" src="https://user-images.githubusercontent.com/63303990/218441842-fd56a1fa-bf5c-4c8a-a7ea-b6f9b5695962.png">

<img width="230" alt="image" src="https://user-images.githubusercontent.com/63303990/218442146-0cc70a03-8234-4a1f-b420-7639f5d24827.png">

Or use the CLI: `cmake -B build -DDEBUG=0 -DGLFW_FETCH=1`
One could then easily mention this clearly in the docs and then voila!
